### PR TITLE
ALTAPPS-570: shared refactoring solved steps flow

### PR DESCRIPTION
**Pull request by vladkash**
**YouTrack Issues**:
[#ALTAPPS-570](https://vyahhi.myjetbrains.com/youtrack/issue/ALTAPPS-570)

**Checklist**

_Before Code Review:_

- [x] Fields "Assignees, Labels, Milestone" are filled in the pull request;
- [x] All checks have been passed;
- [x] Changes have been checked locally.

1. Moved daily study reminders local notification request and daily step completed modal from StepQuiz to StepCompletion
2. Code quiz reset logic is left in StepQuiz
3. Remove RequestUserPermission logic from StepQuiz, because the only permission in StepQuiz is code reset

### CODE REVIEW

**Comment by ivan-magda** [comment]
@vladkash let's always write a small description of what has been done in the PR:
1. Moved daily study reminders local notification request from StepQuiz to StepCompletion
2. ...

cc @XanderZhu

File: shared/src/commonMain/kotlin/org/hyperskill/app/step_completion/presentation/StepCompletionActionDispatcher.kt [hunk 1]
```diff
@@ -28,14 +32,39 @@ class StepCompletionActionDispatcher(
...
+                            .getOrElse { return@onEach }
+
+                        val gemsEarned = currentProfileHypercoinsBalance - cachedProfile.gamification.hypercoinsBalance
+
+                        profileInteractor.notifyHypercoinsBalanceChanged(currentProfileHypercoinsBalance)
+                        onNewMessage(
+                            Message.ShowProblemOfDaySolvedModal(
+                                earnedGemsText = resourceProvider.getQuantityString(SharedResources.plurals.earned_gems, gemsEarned, gemsEarned)
+                            )
+                        )
```
Comment (ivan-magda) [inline shared/src/commonMain/kotlin/org/hyperskill/app/step_completion/presentation/StepCompletionActionDispatcher.kt:63]: Max line length is 120:

```kotlin
onNewMessage(
    Message.ShowProblemOfDaySolvedModal(
        earnedGemsText = resourceProvider.getQuantityString(
            SharedResources.plurals.earned_gems,
            gemsEarned,
            gemsEarned
        )
    )
)
```

File: shared/src/commonMain/kotlin/org/hyperskill/app/step_completion/presentation/StepCompletionReducer.kt [hunk 2]
```diff
@@ -93,11 +117,34 @@ class StepCompletionReducer(private val stepRoute: StepRoute) : StateReducer<Sta
                 )
                 state to setOf(Action.LogAnalyticEvent(event))
             }
-            is Message.TopicCompletedModalHiddenEventMessage -> {
+            Message.TopicCompletedModalHiddenEventMessage -> {
```
Comment (ivan-magda) [inline shared/src/commonMain/kotlin/org/hyperskill/app/step_completion/presentation/StepCompletionReducer.kt:120]: `is Message.TopicCompletedModalHiddenEventMessage`

File: iosHyperskillApp/iosHyperskillApp/Sources/Modules/Step/StepViewModel.swift [hunk 3]
```diff
@@ -73,6 +80,33 @@ final class StepViewModel: FeatureViewModel<StepFeatureState, StepFeatureMessage
...
+
+    func handleSendDailyStudyRemindersPermissionRequestResult(isGranted: Bool) {
+        let message = StepCompletionFeatureMessageRequestDailyStudyRemindersPermissionResult(
+            isGranted: isGranted
+        )
+
+        if isGranted {
+            Task(priority: .userInitiated) {
+                let isNotificationPermissionGranted =
+                await notificationsRegistrationService.requestAuthorizationIfNeeded()
```
Comment (ivan-magda) [inline iosHyperskillApp/iosHyperskillApp/Sources/Modules/Step/StepViewModel.swift:93]: 2 spaces:

```swift
let isNotificationPermissionGranted =
  await notificationsRegistrationService.requestAuthorizationIfNeeded()
```

File: iosHyperskillApp/iosHyperskillApp/Sources/Modules/Step/StepViewModel.swift [hunk 4]
```diff
@@ -123,3 +157,39 @@ extension StepViewModel: TopicCompletedModalViewControllerDelegate {
         )
     }
 }
+
+// MARK: - StepViewModel: ProblemOfDaySolvedModalViewControllerDelegate -
+
+extension StepViewModel: ProblemOfDaySolvedModalViewControllerDelegate {
```
Comment (ivan-magda) [inline iosHyperskillApp/iosHyperskillApp/Sources/Modules/Step/StepViewModel.swift:163]: 👍

File: iosHyperskillApp/iosHyperskillApp/Sources/Modules/Step/Views/Modals/ProblemOfDaySolvedModalViewController.swift [hunk 5]
```diff
@@ -169,7 +179,7 @@ final class ProblemOfDaySolvedModalViewController: PanModalPresentableViewContro
 
     @objc
     private func goBackButtonTapped() {
-        onGoBackButtonTap()
+        delegate?.problemOfDaySolvedModalViewControllerDidTapGoToHomescreenButton(self)
         dismiss(animated: true)
```
Comment (ivan-magda) [inline iosHyperskillApp/iosHyperskillApp/Sources/Modules/Step/Views/Modals/ProblemOfDaySolvedModalViewController.swift:183]: You also dismissing in the `StepViewModel`:

```swift
func problemOfDaySolvedModalViewControllerDidTapGoToHomescreenButton(
 _ viewController: ProblemOfDaySolvedModalViewController
) {
 onNewMessage(
     StepFeatureMessageStepCompletionMessage(
         message: StepCompletionFeatureMessageProblemOfDaySolvedModalGoBackClicked()
     )
 )

 viewController.dismiss(animated: true)
}
```

**Review by ivan-magda — APPROVED** [review]

File: iosHyperskillApp/iosHyperskillApp/Sources/Modules/StepQuiz/StepQuizAssembly.swift [hunk 6]
```diff
@@ -24,8 +24,6 @@ final class StepQuizAssembly: Assembly {
         let commonComponent = AppGraphBridge.sharedAppGraph.commonComponent
```
Comment (ivan-magda) [inline iosHyperskillApp/iosHyperskillApp/Sources/Modules/StepQuiz/StepQuizAssembly.swift:24]: Initialization of value `commonComponent` was never used -> remove it

File: shared/src/commonMain/kotlin/org/hyperskill/app/step_quiz/presentation/StepQuizFeature.kt [hunk 7]
```diff
@@ -69,29 +68,16 @@ interface StepQuizFeature {
...
-        ) : Message
-
-        /**
-         * Show problem of day solve modal
-         */
-
-        data class ShowProblemOfDaySolvedModal(val earnedGemsText: String) : Message
-
-        object ProblemOfDaySolvedModalGoBackClicked : Message
+        object RequestResetCode : Message
```
Comment (XanderZhu) [inline shared/src/commonMain/kotlin/org/hyperskill/app/step_quiz/presentation/StepQuizFeature.kt:73]: Is it assumed to be sent from platform UI?
In my opinion, the name is quite confusing. It is the same as the `Action.ViewAcion.RequestResetCode`.
What about ResetCodeClicked?

Comment (XanderZhu) [reply shared/src/commonMain/kotlin/org/hyperskill/app/step_quiz/presentation/StepQuizFeature.kt:73]: In addition, in my opinion, the name is quite confusing. It is the same as the `Action.ViewAcion.RequestResetCode`.What about ResetCodeClicked?

Comment (ivan-magda) [reply shared/src/commonMain/kotlin/org/hyperskill/app/step_quiz/presentation/StepQuizFeature.kt:73]: `StepQuizFeature.Message.RequestResetCode` is unused -> can be safely deleted

Comment (ivan-magda) [reply shared/src/commonMain/kotlin/org/hyperskill/app/step_quiz/presentation/StepQuizFeature.kt]: We are requesting permission for the code reset when user wants to create new submission (via retry button on Step screen or on fullscreen code editor via reset code button):
```kotlin
is Message.CreateAttemptClicked ->
    if (state is State.AttemptLoaded) {
        if (state.step.block.name == BlockName.CODE || state.step.block.name == BlockName.SQL) {
            state to setOf(Action.ViewAction.RequestResetCode)
        } else {
            State.AttemptLoading(oldState = state) to setOf(
                Action.CreateAttempt(
                    message.step,
                    state.attempt,
                    state.submissionState,
                    message.shouldResetReply
                )
            )
        }
    } else {
        null
    }
```

**Review by XanderZhu — APPROVED** [review]

**Pull request merged**